core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,737 @@
+        T: int. Model parameter 'T' of winnowing.
+        K: int. Model parameter 'K' of winnowing.
+
+    Returns:
+        dict. A dictionary containing training data with additional key
+        'fingerprint' for each program which stores generated fingerprint
+        of program. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                } for each program in training_data.
+            }
+    """
+    window_size = T - K + 1
+    for program_id in training_data:
+        training_data[program_id]['fingerprint'] = (
+            winnowing.get_fingerprint_from_hashes(
+                training_data[program_id]['k_gram_hashes'], window_size))
+    return training_data
+
+
+def calc_jaccard_index(multiset_a, multiset_b):
+    """Calculate jaccard's coefficient for two multisets mutliset_a
+    and multiset_b.
+
+    Jaccard index of two set is equal to:
+        (no. of elements in intersection of two multisets)
+        _____________________________________________
+        (no. of elements in union of two multisets)
+
+    Note: intersection and union of two multisets is similar to union-all and
+    intersect-all operations in SQL.
+
+    Args:
+        multiset_a: list(int). First set.
+        multiset_b: list(int). Second set.
+
+    Returns:
+        float. Jaccard index of two sets.
+
+    """
+    multiset_a = sorted(multiset_a[:])
+    multiset_b = sorted(multiset_b[:])
+    small_set = (
+        multiset_a[:] if len(multiset_a) < len(multiset_b) else multiset_b[:])
+    union_set = (
+        multiset_b[:] if len(multiset_a) < len(multiset_b) else multiset_a[:])
+    index = 0
+    extra_elements = []
+    for elem in small_set:
+        while index < len(union_set) and elem > union_set[index]:
+            index += 1
+        if index >= len(union_set) or elem < union_set[index]:
+            extra_elements.append(elem)
+        elif elem == union_set[index]:
+            index += 1
+
+    union_set.extend(extra_elements)
+    if union_set == []:
+        return 0
+
+    index = 0
+    intersection_set = []
+    for elem in multiset_a:
+        while index < len(multiset_b) and elem > multiset_b[index]:
+            index += 1
+        if index < len(multiset_b) and elem == multiset_b[index]:
+            index += 1
+            intersection_set.append(elem)
+
+    coeff = float(len(intersection_set)) / len(union_set)
+    return coeff
+
+
+def get_program_similarity(fingerprint_a, fingerprint_b):
+    """Find similarity between fingerprint of two programs.
+
+    Args:
+        fingerprint_a: list((int, int)). Fingerprint of first program.
```

Comment:
According to winnowing research paper, `In each window select the minimum hash value. If there is more than one hash with the minimum value, select the rightmost occurrence. Now save all selected hashes as the fingerprints of the document.` This is what winnowing algorithm is. It is can also be used as definition of fingerprint. What do you think I should write in comment to define fingerprint?